SYCL/ESIMD/api/simd_view_copy_move_assign.cpp

```diff
@@ -24,45 +24,56 @@
 using namespace cl::sycl;
 using namespace sycl::ext::intel::experimental::esimd;
 
-template <class T> bool test(queue q, std::string str, T funcUnderTest) {
-  std::cout << "Testing " << str << " ...\n";
-  constexpr unsigned VL = 8;
-
+template <unsigned VL, class T>
+bool test(queue q, std::string str, T funcUnderTest) {
+  std::cout << "Testing " << str << ", VL = " << VL << " ...\n";
   int A[VL];
   int B[VL];
-  // As a result, A should have first 4 values from B, next 4 values from A.
-  int gold[VL] = {0, 1, 2, 3, -4, -5, -6, -7};
+  constexpr unsigned HalfVL = VL > 1 ? (VL / 2) : 1;
 
+  // The expected result gets the first half of values from B,
+  // the rest of values from A.
+  int gold[VL];
   for (unsigned i = 0; i < VL; ++i) {
     A[i] = -i;
     B[i] = i;
+    gold[i] = (i < HalfVL) ? B[i] : A[i];
   }
 
   try {
     buffer<int, 1> bufA(A, range<1>(VL));
     buffer<int, 1> bufB(B, range<1>(VL));
     range<1> glob_range{1};
 
-    auto e = q.submit([&](handler &cgh) {
-      auto PA = bufA.template get_access<access::mode::read_write>(cgh);
-      auto PB = bufB.template get_access<access::mode::read_write>(cgh);
-      cgh.parallel_for(glob_range, [=](id<1> i) SYCL_ESIMD_KERNEL {
-        using namespace sycl::ext::intel::experimental::esimd;
-        unsigned int offset = i * VL * sizeof(int);
-        simd<int, VL> va;
-        va.copy_from(PA, offset);
-        simd<int, VL> vb;
-        vb.copy_from(PB, offset);
-        auto va_view = va.select<4, 1>(0);
-        auto vb_view = vb.select<4, 1>(0);
-
-        funcUnderTest(va_view, vb_view);
-
-        va.copy_to(PA, offset);
-        vb.copy_to(PB, offset);
-      });
-    });
-    q.wait_and_throw();
+    q.submit([&](handler &cgh) {
+       auto PA = bufA.template get_access<access::mode::read_write>(cgh);
+       auto PB = bufB.template get_access<access::mode::read_write>(cgh);
+       cgh.parallel_for(glob_range, [=](id<1> i) SYCL_ESIMD_KERNEL {
+         using namespace sycl::ext::intel::experimental::esimd;
+         unsigned int offset = i * VL * sizeof(int);
+         simd<int, VL> va;
+         simd<int, VL> vb;
+         if constexpr (VL == 1) {
+           va[0] = PA[0];
+           vb[0] = PB[0];
+         } else {
+           va.copy_from(PA, offset);
+           vb.copy_from(PB, offset);
+         }
+
+         auto va_view = va.template select<HalfVL, 1>(0);
+         auto vb_view = vb.template select<HalfVL, 1>(0);
+         funcUnderTest(va_view, vb_view);
+
+         if constexpr (VL == 1) {
+           PA[0] = va[0];
```

Comment:
@kbobrovs - looking at this code, wouldn't you think it may be  a good idea to support copy_to/copy_from for VL=1 by lowering them to such simple assignment? Currently, copy_from/copy_to fail on this assert: https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/ext/intel/experimental/esimd/detail/simd_obj_impl.hpp#L641

Reply:
Yes, VL=1 is a usual simd object, which should support all simd APIs which make any sense. So this assert is a bug.